crates/turbopack-node/src/source_map/mod.rs

```diff
@@ -1,7 +1,371 @@
+                        inside.bold(),
+                        after.dimmed()
+                    )
+                    .unwrap();
+                } else {
+                    writeln!(f, "       | {}{}", " ".repeat(before.len()), "v".bold()).unwrap();
+                    writeln!(
+                        f,
+                        "{line:>6} + {}{}{}",
+                        before.bold(),
+                        inside.bold(),
+                        after.bold()
+                    )
+                    .unwrap();
+                }
+                if inside.len() >= 2 {
+                    writeln!(
+                        f,
+                        "       + {}{}{}{}",
+                        " ".repeat(before.len()),
+                        "^".bold(),
+                        "-".repeat(inside.len() - 2).bold(),
+                        "^".bold(),
+                    )
+                    .unwrap();
+                } else {
+                    writeln!(f, "       | {}{}", " ".repeat(before.len()), "^".bold()).unwrap();
+                }
+            }
+            SourceContextLine::Inside { line, inside } => {
+                writeln!(f, "{:>6} + {}", line.bold(), inside.bold()).unwrap();
+            }
+        }
+    }
+}
+
+enum ResolvedSourceMapping {
+    NoSourceMap,
+    Unmapped,
+    Mapped {
+        frame: StackFrame<'static>,
+    },
+    MappedProject {
+        frame: StackFrame<'static>,
+        project_path: FileSystemPathReadRef,
+        lines: FileLinesContentReadRef,
+    },
+}
+
+async fn resolve_source_mapping(
+    assets_for_source_mapping: AssetsForSourceMappingVc,
+    root: FileSystemPathVc,
+    project_dir: FileSystemPathVc,
+    frame: &StackFrame<'_>,
+) -> Result<ResolvedSourceMapping> {
+    let Some((line, column)) = frame.get_pos() else {
+        return Ok(ResolvedSourceMapping::NoSourceMap);
+    };
+    let name = frame.name.as_ref();
+    let file = &frame.file;
+    let Some(root) = to_sys_path(root).await? else {
+        return Ok(ResolvedSourceMapping::NoSourceMap);
+    };
+    let Ok(file) = Path::new(file.as_ref()).strip_prefix(root) else {
+        return Ok(ResolvedSourceMapping::NoSourceMap);
+    };
+    let file = file.to_string_lossy();
+    let file = if MAIN_SEPARATOR != '/' {
+        Cow::Owned(file.replace(MAIN_SEPARATOR, "/"))
+    } else {
+        file
+    };
+    let map = assets_for_source_mapping.await?;
+    let Some(generate_source_map) = map.get(file.as_ref()) else {
+        return Ok(ResolvedSourceMapping::NoSourceMap);
+    };
+    let Some(sm) = *generate_source_map.generate_source_map().await? else {
+        return Ok(ResolvedSourceMapping::NoSourceMap);
+    };
+    let trace = SourceMapTraceVc::new(sm, line, column, name.map(|s| s.to_string()))
+        .trace()
+        .await?;
+    match &*trace {
+        TraceResult::Found(frame) => {
+            if let Some(project_path) = frame.file.strip_prefix("/turbopack/[project]/") {
```

Comment:
The `"/turbopack/"` prefix is duplicated in three different places. Move it to a constant.

Same comment for `"[project]"`, let's make sure we can't break this code by updating either of these.